src/ws.nim

```diff
@@ -0,0 +1,570 @@
+
+  if cast[int](ws.tcpSocket.fd) == -1:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+
+  # Grab the header.
+  var header: seq[byte]
+  try:
+    header = await ws.tcpSocket.read(2)
+  except:
+    raise newException(WebSocketError, "Socket closed")
+
+  if header.len != 2:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "Socket closed")
+
+  let b0 = header[0].uint8
+  let b1 = header[1].uint8
+
+  # Read the flags and fin from the header.
+  result.fin = b0[0]
+  result.rsv1 = b0[1]
+  result.rsv2 = b0[2]
+  result.rsv3 = b0[3]
+  result.opcode = (b0 and 0x0f).Opcode
+
+  # If any of the rsv are set close the socket.
+  if result.rsv1 or result.rsv2 or result.rsv3:
+    ws.readyState = Closed
+    raise newException(WebSocketError, "WebSocket rsv mismatch")
+
+  # Payload length can be 7 bits, 7+16 bits, or 7+64 bits.
+  var finalLen: uint = 0
+
+  let headerLen = uint(b1 and 0x7f)
+  if headerLen == 0x7e:
+    # Length must be 7+16 bits.
+    var length = await ws.tcpSocket.read(2)
+    if length.len != 2:
+      raise newException(WebSocketError, "Socket closed")
+    finalLen = cast[ptr uint16](length[0].addr)[].htons
+
+  elif headerLen == 0x7f:
+    # Length must be 7+64 bits.
+    var length = await ws.tcpSocket.read(8)
+    if length.len != 8:
+      raise newException(WebSocketError, "Socket closed")
+    finalLen = cast[ptr uint32](length[4].addr)[].htonl
+
+  else:
+    # Length must be 7 bits.
+    finalLen = headerLen
+
+  # Do we need to apply mask?
+  result.mask = (b1 and 0x80) == 0x80
+
+  if ws.masked == result.mask:
+    # Server sends unmasked but accepts only masked.
+    # Client sends masked but accepts only unmasked.
+    raise newException(WebSocketError, "Socket mask mismatch")
+
+  var maskKey: seq[byte]
+  if result.mask:
+    # Read the mask.
+    maskKey = await ws.tcpSocket.read(4)
+    if maskKey.len != 4:
+      raise newException(WebSocketError, "Socket closed")
+
+  # Read the data.
+  var data: seq[byte]
+  data = await ws.tcpSocket.read(int finalLen)
+  result.data = toString(data)
+  if result.data.len != int finalLen:
+    raise newException(WebSocketError, "Socket closed")
+
+  if result.mask:
+    # Apply mask, if we need too.
+    for i in 0 ..< result.data.len:
+      result.data[i] = (result.data[i].uint8 xor maskKey[i mod 4].uint8).char
+
+
+proc receivePacket*(ws: WebSocket): Future[(Opcode, string)] {.async.} =
+  ## Wait for a string or binary packet to come in.
+  var frame = await ws.receiveFrame()
+  result[0] = frame.opcode
+  result[1] = frame.data
+  # If there are more parts read and wait for them
+  while frame.fin != true:
+    frame = await ws.receiveFrame()
+    if frame.opcode != Cont:
+      raise newException(WebSocketError, "Socket closed")
+    result[1].add frame.data
+  return
+
+proc receiveStrPacket*(ws: WebSocket): Future[string] {.async.} =
+  ## Wait only for a string packet to come. Errors out on Binary packets.
```

Comment:
this is a bit confusing, the comment says that this should only handle string packets but it also responds to pings?